Bugfix for validator bond updating

xlab opened the pull request:
New logic from upstream that ensures increased validator bond shares during delegation.

This allows to increase delegation after marking previous delegation as bond, thus allowing to LSM-tokenize more.

Test is in this PR:
https://github.com/persistenceOne/persistenceCore/pull/225

Upstream changes:
https://github.com/iqlusioninc/cosmos-sdk/pull/24

Commit:
fix from upstream: validator bond updating issue

ajeet97 commented on tests/integration/staking/keeper/msg_server_test.go:
```diff
@@ -1355,3 +1355,103 @@ func TestICADelegateUndelegate(t *testing.T) {
+	require.NoError(t, err, "no error expected during redelegation")
+
+	checkValidatorBondShares(validatorAAddress, delegation1Amount.Add(delegation2Amount).Sub(redelegateAmount))
+	checkValidatorBondShares(validatorBAddress, sdk.ZeroInt())
```
Should we flag the delegator as validator bond for B as well?
To check if re-delegating increases bond shares on B.

ajeet97 replied:
it's resolved [here](https://github.com/cosmos/cosmos-sdk/pull/16747#discussion_r1291455437)

xlab replied:
@ajeet97 included those changes now

Commit:
upd with latest upsteam lsm changes, includes fix to redelegation bond accounting

c571075cca2dc8cf1a45c4c0149fbc3566fc344c - increment validator bond shares in cancel unbonding
846943c2596cf87b9026d3389cd739643b0de453 - increased validator bond shares during redelegation
3024452ffacf16158819e0ca92842e05bd38eb50 - removed validator pointer from liquid staking accounting functions
04843f4ab98af92937f585a39e99d28623dc68dc - moved removal of tokenize share locks outside of iterator
8f4c9dc622dcda82c1a065dc5a602a9dcb9598e9 - fixed test
+ 2 new tests

kruspy commented on x/staking/keeper/msg_server.go:
```diff
@@ -300,28 +315,44 @@ func (k msgServer) BeginRedelegate(goCtx context.Context, msg *types.MsgBeginRed
-	if delegation.ValidatorBond {
-		if err := k.SafelyDecreaseValidatorBond(ctx, &srcValidator, srcShares); err != nil {
+	// If this is a validator self-bond, the new liquid delegation cannot fall below the self-bond * bond factor
+	// The delegation on the new validator will not a validator bond
```
typo here ?

xlab replied:
will fix
